Feature/Campaign API integration

brayvasq opened the pull request:
## :loudspeaker: Change type
- [x] New feature

## :scroll: Description
<!--- Describe your changes in detail -->
- [x] Add model for building request body.
- [x] Add service for connecting to CleverTap Campaign API.
- [x] Add new method for creating campaign to `CleverTap` class.
- [x] Split campaigns types into different models.
- [x] Allow to send more than 1000 users per campaign

## :memo: TODO
- [x] Add `CleverTap::Campaign` examples to ReadMe.
- [x] Fix specs for sending `:push` notifications are skipped because of the following error:
  ```json
  {
      "status": "fail",
      "error": "Notification channel is required for devices having Android 8.0 or above",
      "code": 9
  }
  ```

## :green_heart: Testing
Specs

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,126 @@
+    end
+  end
+
+  class NoUsersError < RuntimeError
```
**suggestion**: renombrar a `NoReceiversError` o `NoAddresseesError`

Esas palabras describen mejor la calidad de receptor del usuario en este caso.

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,126 @@
+    MUTABLE_CONTENT = 'mutable-content'.freeze
+    PLATFORM_SPECIFIC = 'platform_specific'.freeze
+
+    def initialize(**args)
```
**suggestion**: usar "Keyword arguments"

Creo que seria mejor apuntar nuestro desarrollo a Ruby 2.0+, y para nosotros vamos a manejar la rama `ruby_2.0.0_support` por separado.
https://thoughtbot.com/blog/ruby-2-keyword-arguments

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,126 @@
+    end
+
+    def to_h
+      put_to_pair
```
**suggestion**: renombrar `put_..._pair` => `..._hash`

2xG replied:
Note que sigues el patron de `CleverTap::Entity` hmmmmm
no me gusta este `put_..._pair`

brayvasq replied:
@2xG no le gusta el nombre de los métodos o la manera en la que se hace el proceso?

2xG commented on README.md:
````diff
@@ -134,6 +134,20 @@ clevertap.upload(event)
+```ruby
+client = CleverTap.new(account_id: '<your account ID>', passcode: '<your passcode>')
+
+campaign = CleverTap::Campaign.new(
````
**nitpick**: usar sintaxis de `CleverTap::Campaign.new \`

porque si, podemos =)

2xG replied:
ay caramba es un ejemplo no mas *facepalm.jpg*

2xG commented on spec/units/campaign_spec.rb:
```diff
@@ -0,0 +1,142 @@
+    subject { described_class.new(**params).to_h }
+
+    let(:params) do
+      {
```
**suggestion**: hacerlo mas compacto

```ruby
{ to: {
    'Email' => ['example@email.com'],
    'FBID' => ['fbidexample']
  },
  tag_group: 'mytaggroup',
  respect_frequency_caps: false,
  content: { 'body' => 'Smsbody' }
}
```

2xG commented on lib/clever_tap/creator.rb:
```diff
@@ -0,0 +1,36 @@
+    TYPE_WEBPUSH = :web_push
+    TYPE_EMAIL = :email
+
+    CAMPAIGNS_NOTIFICATIONS_NAMES = {
```
**suggestion**: renombrar `CAMPAIGNS_NOTIFICATIONS_NAMES` => `CAMPAIGNS_NOTIFICATIONS_ENDPOINTS`

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,126 @@
+    BADGE_ICON = 'badge_icon'.freeze
+    MUTABLE_CONTENT = 'mutable-content'.freeze
+    PLATFORM_SPECIFIC = 'platform_specific'.freeze
+
```
**issue**: agregar `MAX_USERS_PER_CAMPAIGN = 1000`

y crear validacion de cantidad de receptores.

> The to parameter is an object that has the following child properties: FBID, Email, Identity, objectId, which are used as identifiers to target your users. Each request can contain a max of 1000 users across these identifiers.

2xG commented on lib/clever_tap/creator.rb:
```diff
@@ -0,0 +1,36 @@
+
+    private
+
+    def parse_response(http_response)
```
**thought**: aqui podemos pensar en el formato que nos puede servir en nuestro app

brayvasq replied:
@2xG Colocar un yield?

2xG replied:
podemos pensarlo en el momento de utilizar en el app, por ahora dejemoslo asi

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,126 @@
+    end
+
+    def allowed?(indentities)
+      indentities.all? { |identity| ALLOWED_IDENTITIES.include?(identity.to_s) }
```
**suggestion**: encontré una forma elegante para hacer esta prueba

```ruby
(indentities - ALLOWED_IDENTITIES).empty?
```

2xG commented on README.md:
```diff
@@ -134,6 +134,20 @@ clevertap.upload(event)
+  content: { 'body' => "Smsbody" }  
+)
+
+client.create_campaign(campaign)
```
**issue**: falta el metodo `create` para `CleverTap::Client`

es lo que usamos en civico-main https://github.com/PLATAHUB/Civico/blob/prod-civico2-new/config/initializers/clevertap.rb. Y por la existencia de dos aplicaciones de CT que usamos al mismo tiempo no podemos usar la instancia de `CleverTap` y toca configurar dos clientes independientes.

brayvasq committed:
add: Campaign model

brayvasq committed:
add: CleverTap::Creator service
  - add: specs for web_push, sms, email

brayvasq committed:
add: create_campaign to CleverTap::Client
  * add: a few specs for client :push
  * change: some params to keywords params

brayvasq committed:
add: Max users per campaign validation

brayvasq committed:
change: name of some campaign methods

brayvasq commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,152 @@
+        .merge(platform_specific_hash)
+    end
+
+    def to_hash
```
@2xG Este nombre no generaría confusión?

2xG replied:
**suggestion**: `receivers_hash`

2xG commented on README.md:
```diff
@@ -134,6 +134,146 @@ clevertap.upload(event)
+  },
+  platform_specific: {  # Optional
+    'safari' => {
+      'deep_link' => 'https://www.google.com',
```
**nitpick**: para safari - https://apple.com :smile:

2xG commented on README.md:
```diff
@@ -134,6 +134,146 @@ clevertap.upload(event)
+    'chrome' => {
+      'image' => 'https://www.exampleImage.com',
+      'icon' => 'https://www.exampleIcon.com',
+      'deep_link' => 'http://www.example.com',
```
**nitpick**: para chrome - https://google.com :point_left:

2xG commented on README.md:
```diff
@@ -134,6 +134,146 @@ clevertap.upload(event)
+    },
+    'firefox' => {
+      'icon' => 'https://www.exampleIcon.com',
+      'deep_link' => 'https://www.google.com',
```
**nitpick**: para firefox - https://mozilla.org :fire: :fox_face:

2xG commented on README.md:
````diff
@@ -134,6 +134,146 @@ clevertap.upload(event)
 ```
 
+### Create a campaign
+```ruby
````
**suggestion**: agregar link to CT docs

https://developer.clevertap.com/docs/create-campaign-api

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,152 @@
+    end
+  end
+
+  class LimitExceededError < RuntimeError
```
**suggestion**: renombrar `LimitExceededError` -> `ReceiversLimitExceededError`

2xG commented on lib/clever_tap/campaign.rb:
```diff
@@ -0,0 +1,152 @@
+    PLATFORM_SPECIFIC = 'platform_specific'.freeze
+    MAX_USERS_PER_CAMPAIGN = 1000
+
+    def initialize(to: nil,
```
**question**: todos los campos son obligatorios?

brayvasq replied:
@2xG Solamente to: y content:

2xG replied:
**chore**: entonces debe ser asi
```ruby
def initialize(to: , content: , tag_group: nil,
                    ...,
                    platform_specific: nil)
```

2xG commented on lib/clever_tap/creator.rb:
```diff
@@ -0,0 +1,36 @@
+class CleverTap
+  class Creator
```
**suggestion**: `CampaignCreator`

brayvasq committed:
change: improve README
  - change: rename Creator to CampaignCreator.
  - change: rename LimitExceededError to ReceiversLimitExceededError

Commit:
Update README.md

brayvasq committed:
change: make 'a' and 'content' mandatory parameters

brayvasq committed:
Merge pull request #2 from PLATAHUB/spoiler-test

change: Hide large code examples into collapsible blocks

brayvasq committed:
fix: README 'platform_specific' params and specs

brayvasq committed:
add: channel validation when platform_specific has android platform

brayvasq committed:
change: Split campaign types into classes

brayvasq committed:
add: bulk_create_campaign
  add: Campaign 'receivers_limit_exceeded?' method to know if the limit was exceeded
  add: Client 'obtain_receivers' method, to split receivers in campaigns which do not exceed the limit
  add: Client 'bulk_create_campaign' method, allowing to send campaigns with more than 1000 receivers
  change: check if receivers limit was exeeded in the 'create_campaign' method

brayvasq committed:
change: move specific campaigns model to the Campaign scope

brayvasq committed:
change: refactoring the bulk_create_campaign
  change: campaign variable names from 'record' to 'campaign'
  add: empty_receivers? method to validate if there are no receivers
  change: rename bulk_create_campaign to receivers_chunks
  fix: use yield_control expectations

brayvasq committed:
change: remove unnecesary guard condition in the create_campaign method

brayvasq committed:
fix: remove duplicated shared example in 'spec/shared/clever_tap_client.rb'

Commit:
Merge pull request #3 from PLATAHUB/feature/campaigns-classes

Refactoring: Split campaign types into classes

Commit:
Fix readme
Add YARD comments
Small refactoring
Fix code style
Bump gem version
